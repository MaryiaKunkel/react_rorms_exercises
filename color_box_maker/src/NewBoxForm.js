import React, { useState } from "react";

function NewBoxForm({ addBox }) {
  const initialState = {
    color: "",
    width: "",
    height: "",
  };
  const [formData, setFormData] = useState(initialState);

  function handleChange(e) {
    const { name, value } = e.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  }

  function handleSumbit(e) {
    e.preventDefault();
    addBox(formData);
    setFormData(initialState);
  }

  return (
    <form onSubmit={handleSumbit}>
      <label htmlFor="color">Background color:</label>
      <input
        id="color"
        name="color"
        type="text"
        placeholder="e.g. red"
        value={formData.color}
        onChange={handleChange}
      ></input>
      <br />

      <label htmlFor="width">Width:</label>
      <input
        id="width"
        name="width"
        type="text"
        placeholder="e.g. 100px"
        value={formData.width}
        onChange={handleChange}
      ></input>
      <br />

      <label htmlFor="height">Height:</label>
      <input
        id="height"
        name="height"
        type="text"
        placeholder="e.g. 100px"
        value={formData.height}
        onChange={handleChange}
      ></input>
      <br />

      <button className="button">Add box</button>
    </form>
  );
}

export default NewBoxForm;
