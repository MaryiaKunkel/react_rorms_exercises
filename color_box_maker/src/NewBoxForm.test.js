import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm";

it("renders without crashing", function () {
  render(<NewBoxForm />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<NewBoxForm />);
  expect(asFragment()).toMatchSnapshot();
});

it("should update the form data state according to the changes", function () {
  const { getByLabelText } = render(<NewBoxForm />);
  const colorInput = getByLabelText("Background color:");
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");

  fireEvent.change(colorInput, { target: { value: "red" } });
  fireEvent.change(widthInput, { target: { value: "150px" } });
  fireEvent.change(heightInput, { target: { value: "250px" } });

  expect(colorInput.value).toBe("red");
  expect(widthInput.value).toBe("150px");
  expect(heightInput.value).toBe("250px");
});

it("should call addBox function when sumbit", function () {
  const addBoxMock = jest.fn();
  const { getByLabelText } = render(<NewBoxForm addBox={addBoxMock} />);
  const colorInput = getByLabelText("Background color:");
  const widthInput = getByLabelText("Width:");
  const heightInput = getByLabelText("Height:");
  const button = getByClassName("button");

  fireEvent.change(colorInput, { target: { value: "red" } });
  fireEvent.change(widthInput, { target: { value: "150px" } });
  fireEvent.change(heightInput, { target: { value: "250px" } });
  fireEvent.click(button);

  expect(addBoxMock).toHaveBeenCalledWith({
    color: "red",
    width: "100px",
    height: "100px",
  });
});
