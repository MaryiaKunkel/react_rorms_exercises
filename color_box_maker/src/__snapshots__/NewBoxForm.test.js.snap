// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="color"
    >
      Background color:
    </label>
    <input
      id="color"
      name="color"
      placeholder="e.g. red"
      type="text"
      value=""
    />
    <br />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      placeholder="e.g. 100px"
      type="text"
      value=""
    />
    <br />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      placeholder="e.g. 100px"
      type="text"
      value=""
    />
    <br />
    <button
      class="button"
    >
      Add box
    </button>
  </form>
</DocumentFragment>
`;
